### PR TITLE
[SQL][SPARK-17439] Fixing compression issues with approximate quantiles and adding more tests

### PR DESCRIPTION
**Pull request by thunterdb**
## What changes were proposed in this pull request?

This PR build on #14976 and fixes a correctness bug that would cause the wrong quantile to be returned for small target errors.
## How was this patch tested?

This PR adds 8 unit tests that were failing without the fix.

### CODE REVIEW

**Comment by thunterdb** [comment]
cc @clockfly

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala [hunk 1]
```diff
@@ -258,7 +263,10 @@ object QuantileSummaries {
     }
     res.prepend(head)
     // If necessary, add the minimum element:
-    res.prepend(currentSamples.head)
+    val currHead = currentSamples.head
+    if (currHead.value < head.value) {
```
Comment (thunterdb) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala]: this is the fix

**Comment by srowen** [comment]
This seems fine to me, and would also fix SPARK-17306. You can note that in the title. I'm closing https://github.com/apache/spark/pull/14976

**Comment by srowen** [comment]
Merged to master and to 2.0 with some conflict resolution

**Comment by thunterdb** [comment]
Thank you @srowen

**Pull request closed without merging**